keps/draft-repo.md

```diff
@@ -0,0 +1,108 @@
+---
+kep-number: draft-15
+title: repo
+authors:
+  - "@kensipe"
+owners:
+  - "@kensipe"
+  - "@fabianbaier"
+creation-date: 2019-07-24
+last-updated: 2019-07-24
+status: draft
+---
+
+# Repository Management
+
+## Table of Contents
+
+* [Repository Management](#repository-management)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Create Operator Tarball](#create-operator-tarball)
+         * [Create Repository Index](#create-repository-index)
+         * [Create Local Respository](#create-local-respository)
+         * [Update a Repository](#update-a-repository)
+         * [CLI Repository Help](#cli-repository-help)
+         * [User Stories](#user-stories)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+There is a number of ways that an operator can be installed via KUDO.  The operator developer can install via the local file system (with a operator folder or tarball).  It is also possible to install via url to a tarball.  While [KEP-0010](keps/0010-package-manager.md) defines the packaging structure, this kep defines:
+
+* The structure of a repository
+* Defining authoritative repositories: incubator & graduated
+* Working with multiple repositories
+* Packaging and Repository management tooling
+* Configuration of repositories for the CLI
+* CLI Tooling for search, packaging, and pushing to/from repository
+* Define the rules and means to promote incubated operators to the graduated repository.
+
+By providing a repository (or set of repositories), KUDO will provide a mechanism for administrators to easy discover and install KUDO operators.  [Helm has provided some prior art](https://github.com/helm/helm/blob/master/docs/chart_repository.md) which we should align closely to from a mental model stand point.
```

Comment:
So.. while we're taking a note from Helm's book.. Helm is actually moving away from the custom chart repository model: https://blog.bacongobbler.com/post/2019-01-25-distributing-with-distribution/index.html

So they outline a number of problems with the custom chart repository model:

> While there are several benefits to the Chart Repository API meeting the most basic storage requirements, a few drawbacks have started to show:
> * Chart Repositories have a very hard time abstracting most of the security implementations required in a production environment. Having a standard API for authentication and authorization is very important in production scenarios
> * Helm’s Chart provenance tools used for signing and verifying the integrity and origin of a chart is an after-thought of the chart publishing process
> * In multi-tenant scenarios, the same chart can be uploaded by another tenant, costing twice the storage cost to store the same content. Smarter chart repositories can be designed to handle this, but it’s not a part of the spec
> * Using a single index file for search, metadata information and fetching charts has made it difficult or clunky to design around in secure multi-tenant implementations.

Compared to using OCI-compatible registries for a backend:

> Let’s talk about what the Distribution project brings to the table:
> * A stable, battle-hardened product with many cloud vendors providing their own flavour of Distribution
> * A rich set of libraries for interacting with Distribution
> * A pluggable storage backend, including support for local filesystem, Azure, GCS, Aliyun OSS, AWS S3, and Openstack Swift out of the box
> * A standard API for authentication/authorization
> * Content stored in Distribution is content-addressable via a sha256 checksum. In other words, if two files are uploaded with identical content, only one copy of the content is stored.

So it seems like quite a bit of the work is already done for us in the OCI image format and repositories. Maybe we should explore that more?

Reply:
totally

Reply:
Oh niiiice, why haven't I seen this before!! :) Couple of thoughts:
- I think our current model is fine up until the point where we want to support people running their own registries (eg. airgaps? other use cases?) then we basically run to the same "problems" as they note there
- the downside of using repository model is that you actually have to run the repository 'application' somewhere and manage it while right now we just use hosted GCS and don't need to maintain official repo app that has to be up all the time

all and all I think we should investigate both options and probably write pros and cons to this list. Maybe we can stick with current model for now and move to that one later? (it worked for helm for quite some time) But maybe that does not make sense if we would invest a lot of time in the current model that will just get thrown away.

Also not sure what that means for the V2 deliverable next friday :)